### ISSUE TITLE
Phpactor not recognising standard php functions

### ISSUE THREAD

**Issue by bchubb-web** [phpactor/phpactor#2754]
I use the phpactor lsp inside of neovim, and codebase-wise it works perfectly, i much prefer it to intelephense, but recently it has stopped recognising all standard php functions, i experienced this with intelephense when i used stubs to add wordpress functions, but i have no special config in my lsp settings, does anyone know how to resolve this?

**Comment by GosuTeacher** [comment]
I've taken a bit of time to check this out as it was bugging me also and found out that after the PHAR indexing the path scheme logic is changed and generates `file:///` scheme instead of `phar://` for the stubs. (When the project is loaded from phar, the phar is somewhere in the middle and the scheme remains as a file, the uri results in `file:///` and it can't be loaded)

From what I've checked, TextDocumentLocator works like this:
1. scheme is already set to one of the supported (phar, file) (not the case for the jetbrains stubs as they are set as canonical absolute path)
2. checks if the path ends with `phar`(extension) to set the scheme for the path  (not the case for the jetbrains stubs as they are set as canonical absolute path)

The second issue that impacts this is that the project is delivered in a `phar` file, if you unwrap the `phpactor.phar` and change the bash file that executes the app, or short-circuit it to bin/phpactor directly, this will work

So, we have a few solutions for this:
1. The easiest one and I think the one that's recommended until this issue is fixed is to change the `https://phpactor.readthedocs.io/en/master/reference/configuration.html#worse-reflection-stub-dir` stub dir config (either at the system/user/project level) like this: `https://phpactor.readthedocs.io/en/master/usage/configuration.html` either to the stubs delivered with the phpactor project, or you can install the `jetbrains/phpstorm-stubs` via composer and point to that location(I'll try to create a PoC for this today)
2. Change the scheme for the config in the app entry point based on how you interact with phpactor (via phar or bin/phpactor) - check if `$phpactorBin` is a phar archive or not. In my opinion this is the worse way to fix it
3. Change the way the scheme is parsed in `TextDocumentLocator`, but this may not be so reliable. e.g. check if the path contains the `.phar/` string and set the scheme as phar

Hope this helps. Good luck!

**Comment by charjr** [comment]
Thank you @GosuTeacher for the clarification, and for the temporary solution!

This issue is still present in the latest release: [2024.11.05.0](https://github.com/phpactor/phpactor/releases/tag/2024.11.05.0)

**Comment by dantleech** [comment]
fwiw I looked into this last night and it seems it should work in theory, but somewhere the `phar://` scheme is getting lost from the stubs during the indexing process, so they are all skipped. will investigate further.

**Comment by GosuTeacher** [comment]
I don't think this generates a `phar://` schemed uri: https://github.com/phpactor/phpactor/blob/44b0af9f4c06116962d1c57ab4535a7a4997f384/lib/Phpactor.php#L112

**Comment by dantleech** [comment]
I thought I checked that yesterday and it did, but will check again, thanks :)

**Comment by GosuTeacher** [comment]
I tested this by repacking the phar with a var dump after that line and running it :))

**Comment by dantleech** [comment]
Yep me too!

**Comment by dantleech** [comment]
found it: https://github.com/phpactor/phpactor/pull/2774

**Comment by dantleech** [comment]
would be great if somebody could confirm that it works although I realise that means building the PHAR otherwise I'll tag a new release in the next days.

**Comment by GosuTeacher** [comment]
Nice, I'll try to do it in a couple of hours. Thank you!

**Comment by GosuTeacher** [comment]
```
Error when executing textDocument/definition : Exception [Phpactor\TextDocument\Exception\TextDocumentNotFound] Text document "file:///home/<user>/.local/share/nvim/mason/packages/phpactor/phpactor-build/build/phpactor.phar/vendor/jetbrains/phpstorm-stubs/standard/standard_7.php" not found at phar:///home/<user>/.local/share/nvim/mason/packages/phpactor/phpactor-build/build/phpactor.phar/lib/TextDocument/Exception/TextDocumentNotFound.php#12
```

Same same :(

**Comment by dantleech** [comment]
I mean, that's _expected_ you can't go to the definition in the PHAR, but autocompletion and hover should work

**Comment by GosuTeacher** [comment]
Those worked before. If you're referring to the function arguments autocomplete, not function suggestions. 
I think the case for this post was go to definition, at least that's what I understood and  investigated. Before 06.30 this worked, and If you pass the stubs via the config(even in PHAR), that also works.

If that's not the case or if it's not in scope, then there's no problem.

**Comment by dantleech** [comment]
VIM and, as far as I know, VS code are not capable of opening a file in a PHAR file, so goto definition would never have worked unless you had your stubs in another location on the filesystem.

What is the case for going to the definition of a stub file?

**Comment by GosuTeacher** [comment]
Setting this path in Neovim worked for me. Maybe VIM/VSCode don't. I'm kinda new to the LSP/editor part of things. In my ignorance, the first thing that came to my mind after I went into using an LSP was go to definition and I used this quite a lot (I didn't know about the hover feature until much later). I was doing this before in PHPStorm. After the release of 06.30 this broke and now we're here.
From what I understand this was not intended to work in the first place.

```
{
	"worse_reflection.stub_dir": "phar://home/<user>/.local/share/nvim/mason/packages/phpactor/phpactor.phar/vendor/jetbrains/phpstorm-stubs"
}
```

**Comment by dantleech** [comment]
I don't know how that worked, as I said VIM can't open PHAR files (unless you have some extension that can), it's possible you had the phpstorm-stubs in he project you were wokring on as a dev-dependency.

**Comment by MaG21** [comment]
@GosuTeacher Thank you for your input.

For anyone interested:

### Solution
TL;DR
```bash
composer require --dev jetbrains/phpstorm-stubs
```

Why does this work ?
Because by default `phpactor` points the configuration `worse_reflection.stub_dir` to `%application_root%/vendor/jetbrains/phpstorm-stubs`. By installing phpstorm stubs you get all of that.

Feel free to check the documentation for `worse_reflection.stub_dir` here: https://phpactor.readthedocs.io/en/master/reference/configuration.html#worse-reflection-stub-dir

**Comment by MaG21** [comment]
For anyone using phpstorm-stubs, may be wise to allow phpactor some time to scan and to close and open your vim again.

**Comment by TRPB** [comment]
Definitely something has changed here, I definitely didn't have the stubs installed via composer previously. Is there a way to force phpactor to rebuild the cache? I've updated `worse_reflection.stub_dir` and verified it's set in `:PhpactorConfig` in nvim but it's still giving me method not found for all the inbuilt methods

**Comment by dantleech** [comment]
`phpactor index:build --reset` in your projects root dir

**Comment by danrot** [comment]
I've been trying working with the new release, and it seems to be partially working... Not sure what the issue remains though.

E.g. I don't get autocompletion for the `explode` function:

![image](https://github.com/user-attachments/assets/f00df1d3-28fd-4573-9fa2-f49f5d79e86b)

But when I type it out manually, it seems to know about the parameters of the `explode` function:

![image](https://github.com/user-attachments/assets/04c0c0e6-f4c7-4e75-984c-21a1de9357d0)

I cannot say for sure though if that was working before the update like that as well...

**Comment by dantleech** [comment]
I'm assuming:

```
phpactor index:search --short-name-begins explode
function # explode
```
returns nothing for you?

**Comment by danrot** [comment]
@dantleech It executes successfully with a exit code of `0`, but there is no output.

**Comment by dantleech** [comment]
so that means that it's still not indexed

**Comment by danrot** [comment]
How does it know about the parameters then? 🤔

**Comment by dantleech** [comment]
hopefully this release fixes it: https://github.com/phpactor/phpactor/releases/tag/2024.11.28.1 let me know otherwise...

**Comment by danrot** [comment]
@dantleech Thank you, it is working now 🙂 Except for jump to definition, but if I understood that correctly before that cannot work with the PHAR, so that's probably fine.

**Comment by dantleech** [comment]
🎉

**Comment by boonkerz** [comment]
Version: 2024.11.28.1
it does not work or?

```thomas@DESKTOP-E17406D:~/.local/share/nvim/mason/bin$ phpactor index:search --short-name-begins explode -vvv

Box Requirements Checker
========================

> Using PHP 8.1.2-1ubuntu2.20
> PHP is using the following php.ini file:
  /etc/php/8.1/cli/php.ini

> Checking Box requirements:
  ✔ The application requires a version matching "^8.1".
  ✔ The application requires the extension "zlib".
  ✔ The package "amphp/dns" requires the extension "filter".
  ✔ The package "amphp/dns" requires the extension "json".
  ✔ The package "ramsey/uuid" requires the extension "json".
  ✔ The package "amphp/socket" requires the extension "openssl".
  ✔ The package "kelunik/certificate" requires the extension "openssl".
  ✔ The package "symfony/service-contracts" conflicts with the extension "psr".


 [OK] Your system is ready to run the application.


[FPR    ][DEBU][803400.810810] Resolved path "%cache%/index/%project_id%" to "/home/thomas/.cache/phpactor/index/bin-d0e07c"
[FPR    ][DEBU][803400.811864] Resolved path "%project_root%" to "/home/thomas/.local/share/nvim/mason/bin"
[FPR    ][DEBU][803400.812629] Resolved path "%project_root%" to "/home/thomas/.local/share/nvim/mason/bin"
[FPR    ][DEBU][803400.827333] Resolved path "/home/thomas/projekte/phpstorm-stubs" to "/home/thomas/projekte/phpstorm-stubs"
[FPR    ][DEBU][803400.827461] Resolved path "%cache%/worse-reflection" to "/home/thomas/.cache/phpactor/worse-reflection"
[FPR    ][DEBU][803400.828147] Resolved path "%project_root%/vendor/autoload.php" to "/home/thomas/.local/share/nvim/mason/bin/vendor/autoload.php"
[COMPOSE][WARN][803400.828259] Could not find autoloader "/home/thomas/.local/share/nvim/mason/bin/vendor/autoload.php"
[FPR    ][DEBU][803400.828298] Resolved path "%project_root%" to "/home/thomas/.local/share/nvim/mason/bin"
thomas@DESKTOP-E17406D:~/.local/share/nvim/mason/bin$ phpactor index:search --short-name-begins explode
thomas@DESKTOP-E17406D:~/.local/share/nvim/mason/bin$ phpactor index:build --reset
Building job...done
Building index:

 518/518 [============================] 100% < 1 sec/< 1 sec 28/∞ mb

Done in 0.11 seconds using 27,648,000b of memory
thomas@DESKTOP-E17406D:~/.local/share/nvim/mason/bin$ phpactor index:search --short-name-begins explode
thomas@DESKTOP-E17406D:~/.local/share/nvim/mason/bin$
```

**Comment by bchubb-web** [comment]
@dantleech This was resolved for a while but is now an issue again, after re-reading this thread i understand it slightly more

all local functions/from vendor are fine, just spl functions, when typing them out i get recommendations in my autocomplete for the correct function, but i dont get autocomplete/prompts for the parameter names, this seems to be the opposite of danrot's issue

all standard functions in a file say they aren't defined (not found), but when using index:search they are found

This is only an issue on a single repo, which is very strange, its quite large (1000000 lines of php) but i doubt that's the issue

**Comment by Taluu** [comment]
Hey, 

Sorry to bump up the subject but I think I'm also getting a strange behavior. I have tried to use the `composer global require ...` for the phpstubs, and some functions are recognized while some others are not. See the screenshot :

![Image](https://github.com/user-attachments/assets/6898dd9e-9226-4b7d-b533-151a27d12d28)

You can see it acknowledges the `is_writable` function no problem, but not the `file` function. Even with the `phpactor index:search --short-name file` (or even `explode` in the example you gave earlier), it doesn't show anything (except a class for `file`, but if I add `--is-function` to the search query, nothing is returned).

I checked in the stubs that these are declared, and they actually are... I'm currently using the phar, but my config does show I have the correct path for the `subs` as suggested a few times in this issue.

I do not have such issues with intelephense though (had to dig up the license key I had made a few years ago :D) but things are bothering me with it, that phpactor brought up and would like to keep. :)

**Comment by CaptainQuirk** [comment]
I confirm that this is happening to me too.
The code base is also quite large.

I get hover information on various built-in php functions, like json_encode, but the name is not resolved, showing an error in the sign column. Now, it's a bit confusing, especially on large files where lots of errors are reported : I struggle to actually discriminate from actual errors and those.

Autocompletion for those functions doesn't work either !

I'm on the master branch, commit `a0bd6074`, on vim 9.1 on WSL and PHP 8.3
